src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java

```diff
@@ -119,12 +121,14 @@ public void failure(Throwable error) {
 
     public static void failure(String k, Throwable error) {
         State s = State.get();
-        if (s.contexts.containsKey(k)) {
-            System.err.println("Unblocking " + k + " as failure");
-            getContext(s, k).onFailure(error);
-        } else {
-            System.err.println("Planning to unblock " + k + " as failure");
-            s.errors.put(k, error);
+        synchronized (s) {
```

Comment:
I did not notice some tests that failed because of the race condition here, but since we are synchronizing the success method, then we need to synchronize here too.